In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import os

In [3]:
os.chdir('data')

In [4]:
os.listdir()

['.ipynb_checkpoints',
 'Calcultations',
 'Cleaned csv',
 'combined_ressources_redlist.csv',
 'Datas  on population since 1990',
 'Environmental Sustainability',
 'hdro_statistical_data_tables_1_15_d1_d5.xlsx',
 'Human Development Index (HDI) 1990.csv',
 'Human Development Index (HDI).csv']

In [6]:
hdi=pd.read_csv('Human Development Index (HDI).csv')

In [7]:
hdi.head()

HDI Rank       Country   1990  \
0  Human Development Index (HDI)\nnull\nDimension...           NaN    NaN   
1                                                169   Afghanistan  0.302   
2                                                 69       Albania  0.650   
3                                                 91       Algeria  0.572   
4                                                 36       Andorra     ..   

   Unnamed: 3   1991  Unnamed: 5   1992  Unnamed: 7   1993  Unnamed: 9  ...  \
0         NaN    NaN         NaN    NaN         NaN    NaN         NaN  ...   
1         NaN  0.307         NaN  0.316         NaN  0.312         NaN  ...   
2         NaN  0.631         NaN  0.615         NaN  0.618         NaN  ...   
3         NaN  0.576         NaN  0.582         NaN  0.586         NaN  ...   
4         NaN     ..         NaN     ..         NaN     ..         NaN  ...   

    2015  Unnamed: 53   2016  Unnamed: 55   2017  Unnamed: 57   2018  \
0    NaN          NaN    NaN          NaN    NaN          NaN    NaN   
1  0.500          NaN  0.502          NaN  0.506          NaN  0.509   
2  0.788          NaN  0.788          NaN  0.790          NaN  0.792   
3  0.740          NaN  0.743          NaN  0.745          NaN  0.746   
4  0.862          NaN  0.866          NaN  0.863          NaN  0.867   

   Unnamed: 59   2019  Unnamed: 61  
0          NaN    NaN          NaN  
1          NaN  0.511          NaN  
2          NaN  0.795          NaN  
3          NaN  0.748          NaN  
4          NaN  0.868          NaN  

[5 rows x 62 columns]

Drop all the columns unnamed which come from the csv. 

In [8]:
hdi.drop(['Unnamed: 3', 'Unnamed: 5', 'Unnamed: 7', 'Unnamed: 9', 'Unnamed: 11', 'Unnamed: 13', 
              'Unnamed: 15', 'Unnamed: 17', 'Unnamed: 19', 'Unnamed: 21', 'Unnamed: 23', 'Unnamed: 25',
              'Unnamed: 27', 'Unnamed: 29', 'Unnamed: 31', 'Unnamed: 33', 'Unnamed: 35', 'Unnamed: 27',
               'Unnamed: 29', 'Unnamed: 31', 'Unnamed: 33', 'Unnamed: 35', 'Unnamed: 37', 'Unnamed: 39',
               'Unnamed: 41', 'Unnamed: 43', 'Unnamed: 45', 'Unnamed: 47', 'Unnamed: 49', 'Unnamed: 51',
               'Unnamed: 53', 'Unnamed: 55','Unnamed: 57', 'Unnamed: 59', 'Unnamed: 61'], axis=1, inplace=True)

In [9]:
hdi

HDI Rank  \
0    Human Development Index (HDI)\nnull\nDimension...   
1                                                  169   
2                                                   69   
3                                                   91   
4                                                   36   
..                                                 ...   
203                                                      
204                                                      
205                                                      
206                                                      
207                                                NaN   

                                               Country   1990   1991   1992  \
0                                                  NaN    NaN    NaN    NaN   
1                                          Afghanistan  0.302  0.307  0.316   
2                                              Albania  0.650  0.631  0.615   
3                                              Algeria  0.572  0.576  0.582   
4                                              Andorra     ..     ..     ..   
..                                                 ...    ...    ...    ...   
203                          Least Developed Countries  0.350  0.353  0.354   
204                     Small Island Developing States  0.595  0.598  0.603   
205  Organization for Economic Co-operation and Dev...  0.785  0.790  0.788   
206                                              World  0.598  0.601  0.601   
207                                                NaN    NaN    NaN    NaN   

      1993   1994   1995   1996   1997  ...   2010   2011   2012   2013  \
0      NaN    NaN    NaN    NaN    NaN  ...    NaN    NaN    NaN    NaN   
1    0.312  0.307  0.331  0.335  0.339  ...  0.472  0.477  0.489  0.496   
2    0.618  0.624  0.637  0.646  0.645  ...  0.745  0.764  0.775  0.782   
3    0.586  0.590  0.595  0.602  0.611  ...  0.721  0.728  0.728  0.729   
4       ..     ..     ..     ..     ..  ...  0.837  0.836  0.858  0.856   
..     ...    ...    ...    ...    ...  ...    ...    ...    ...    ...   
203  0.358  0.358  0.366  0.374  0.381  ...  0.485  0.493  0.499  0.504   
204  0.608  0.612  0.618  0.624  0.629  ...  0.702  0.706  0.704  0.708   
205  0.800  0.807  0.812  0.817  0.817  ...  0.873  0.877  0.879  0.883   
206  0.608  0.611  0.617  0.622  0.624  ...  0.697  0.703  0.708  0.713   
207    NaN    NaN    NaN    NaN    NaN  ...    NaN    NaN    NaN    NaN   

      2014   2015   2016   2017   2018   2019  
0      NaN    NaN    NaN    NaN    NaN    NaN  
1    0.500  0.500  0.502  0.506  0.509  0.511  
2    0.787  0.788  0.788  0.790  0.792  0.795  
3    0.736  0.740  0.743  0.745  0.746  0.748  
4    0.863  0.862  0.866  0.863  0.867  0.868  
..     ...    ...    ...    ...    ...    ...  
203  0.510  0.516  0.520  0.525  0.528  0.538  
204  0.712  0.717  0.719  0.722  0.723  0.728  
205  0.886  0.889  0.892  0.894  0.895  0.900  
206  0.718  0.722  0.727  0.729  0.731  0.737  
207    NaN    NaN    NaN    NaN    NaN    NaN  

[208 rows x 32 columns]

In [10]:
hdi.columns

Index(['HDI Rank', 'Country', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018', '2019'],
      dtype='object')

In [11]:
hdi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 32 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   HDI Rank  205 non-null    object
 1   Country   206 non-null    object
 2   1990      206 non-null    object
 3   1991      206 non-null    object
 4   1992      206 non-null    object
 5   1993      206 non-null    object
 6   1994      206 non-null    object
 7   1995      206 non-null    object
 8   1996      206 non-null    object
 9   1997      206 non-null    object
 10  1998      206 non-null    object
 11  1999      206 non-null    object
 12  2000      206 non-null    object
 13  2001      206 non-null    object
 14  2002      206 non-null    object
 15  2003      206 non-null    object
 16  2004      206 non-null    object
 17  2005      206 non-null    object
 18  2006      206 non-null    object
 19  2007      206 non-null    object
 20  2008      206 non-null    object
 21  2009      206 no

In [12]:
hdi.duplicated().value_counts()

False    208
dtype: int64

# Data Cleaning

## First analysis

- The values in the dataframe must be slightly changed.
1. The column HDI Rank can be dropped, to avoid problem during analysis.
2. Problem with the values, because it's objects and not float. 
3. The HDI rank must be in the last place: it's not relevant for the moment.
4. Some values doesn't exist, so it must be replaced by NaN.

To avoid problems with the DataFrame if there is a problem, a backup copy has to be done.

In [13]:
df = hdi.copy()

In [14]:
# Drop row HDI Rank
df.drop('HDI Rank', axis=1, inplace=True)

In [15]:
# Replace the .. which are aliases for no value and spaces 
df=df.replace('..', '').replace(' ', '')

In [16]:
# Select the columns to transform it into float excluding the Country column
cols = df.columns[df.dtypes.eq('object')]
cols=cols.drop('Country')

In [17]:
df[cols]=df[cols].apply(pd.to_numeric, downcast='float')
df.head()

Country   1990   1991   1992   1993   1994   1995   1996   1997  \
0           NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
1   Afghanistan  0.302  0.307  0.316  0.312  0.307  0.331  0.335  0.339   
2       Albania  0.650  0.631  0.615  0.618  0.624  0.637  0.646  0.645   
3       Algeria  0.572  0.576  0.582  0.586  0.590  0.595  0.602  0.611   
4       Andorra    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

    1998  ...   2010   2011   2012   2013   2014   2015   2016   2017   2018  \
0    NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
1  0.344  ...  0.472  0.477  0.489  0.496  0.500  0.500  0.502  0.506  0.509   
2  0.655  ...  0.745  0.764  0.775  0.782  0.787  0.788  0.788  0.790  0.792   
3  0.621  ...  0.721  0.728  0.728  0.729  0.736  0.740  0.743  0.745  0.746   
4    NaN  ...  0.837  0.836  0.858  0.856  0.863  0.862  0.866  0.863  0.867   

    2019  
0    NaN  
1  0.511  
2  0.795  
3  0.748  
4  0.868  

[5 rows x 31 columns]

## Checking the NaN values

In [18]:
df.isna().sum()

Country     2
1990       49
1991       49
1992       50
1993       49
1994       49
1995       45
1996       45
1997       46
1998       45
1999       42
2000       19
2001       19
2002       18
2003       17
2004       15
2005        8
2006        7
2007        7
2008        7
2009        7
2010        5
2011        5
2012        5
2013        5
2014        5
2015        5
2016        5
2017        4
2018        4
2019        4
dtype: int64

If the columns seems to be okay, let see if there is the same for the rows.

In [19]:
for i in range(len(df.index)) :
    print("Nan in row ", i , " : " ,  df.iloc[i].isna().sum()/df.shape[1])

Nan in row  0  :  1.0
Nan in row  1  :  0.0
Nan in row  2  :  0.0
Nan in row  3  :  0.0
Nan in row  4  :  0.3225806451612903
Nan in row  5  :  0.2903225806451613
Nan in row  6  :  0.4838709677419355
Nan in row  7  :  0.0
Nan in row  8  :  0.0
Nan in row  9  :  0.0
Nan in row  10  :  0.0
Nan in row  11  :  0.16129032258064516
Nan in row  12  :  0.3225806451612903
Nan in row  13  :  0.0
Nan in row  14  :  0.0
Nan in row  15  :  0.0
Nan in row  16  :  0.16129032258064516
Nan in row  17  :  0.0
Nan in row  18  :  0.0
Nan in row  19  :  0.0
Nan in row  20  :  0.4838709677419355
Nan in row  21  :  0.0
Nan in row  22  :  0.3225806451612903
Nan in row  23  :  0.0
Nan in row  24  :  0.0
Nan in row  25  :  0.0
Nan in row  26  :  0.0
Nan in row  27  :  0.3225806451612903
Nan in row  28  :  0.0
Nan in row  29  :  0.3225806451612903
Nan in row  30  :  0.0
Nan in row  31  :  0.0
Nan in row  32  :  0.0
Nan in row  33  :  0.0
Nan in row  34  :  0.3225806451612903
Nan in row  35  :  0.0
Nan in row  36 

It seems to be okay for most of the rows, but 2 have more that 90% of missing values : rows `189` and `195`.
Rows `108`, `158` and `175` have a missing values rate over 60 %. Let's have a look to those rows.

In [20]:
# What is in the bottom of the DataFrame?
df.tail(20)

Country   1990   1991   1992  \
188                                             Zambia  0.421  0.417  0.416   
189                                           Zimbabwe  0.478  0.481  0.467   
190                                  Human Development    NaN    NaN    NaN   
191                        Very high human development  0.779  0.782  0.780   
192                             High human development  0.568  0.573  0.578   
193                           Medium human development  0.437  0.439  0.445   
194                              Low human development  0.352  0.353  0.355   
195                               Developing Countries  0.516  0.520  0.525   
196                                            Regions    NaN    NaN    NaN   
197                                        Arab States  0.556  0.559  0.565   
198                          East Asia and the Pacific  0.519  0.525  0.534   
199                            Europe and Central Asia  0.652  0.650  0.646   
200                    Latin America and the Caribbean  0.628  0.634  0.639   
201                                         South Asia  0.441  0.447  0.454   
202                                 Sub-Saharan Africa  0.402  0.404  0.404   
203                          Least Developed Countries  0.350  0.353  0.354   
204                     Small Island Developing States  0.595  0.598  0.603   
205  Organization for Economic Co-operation and Dev...  0.785  0.790  0.788   
206                                              World  0.598  0.601  0.601   
207                                                NaN    NaN    NaN    NaN   

      1993   1994   1995   1996   1997   1998  ...   2010   2011   2012  \
188  0.419  0.414  0.415  0.416  0.416  0.416  ...  0.527  0.534  0.549   
189  0.463  0.460  0.453  0.453  0.447  0.442  ...  0.482  0.499  0.525   
190    NaN    NaN    NaN    NaN    NaN    NaN  ...    NaN    NaN    NaN   
191  0.790  0.795  0.799  0.804  0.804  0.813  ...  0.866  0.871  0.874   
192  0.584  0.588  0.596  0.604  0.610  0.616  ...  0.706  0.713  0.720   
193  0.451  0.457  0.464  0.471  0.476  0.484  ...  0.575  0.584  0.593   
194  0.356  0.356  0.361  0.368  0.373  0.379  ...  0.473  0.479  0.484   
195  0.530  0.534  0.541  0.547  0.553  0.559  ...  0.642  0.650  0.657   
196    NaN    NaN    NaN    NaN    NaN    NaN  ...    NaN    NaN    NaN   
197  0.571  0.576  0.581  0.587  0.594  0.599  ...  0.676  0.681  0.687   
198  0.542  0.549  0.559  0.568  0.576  0.581  ...  0.691  0.700  0.707   
199  0.643  0.638  0.642  0.646  0.651  0.657  ...  0.735  0.744  0.750   
200  0.644  0.650  0.656  0.662  0.669  0.675  ...  0.731  0.737  0.740   
201  0.460  0.466  0.473  0.480  0.486  0.492  ...  0.585  0.593  0.601   
202  0.406  0.407  0.411  0.415  0.418  0.422  ...  0.498  0.505  0.512   
203  0.358  0.358  0.366  0.374  0.381  0.388  ...  0.485  0.493  0.499   
204  0.608  0.612  0.618  0.624  0.629  0.632  ...  0.702  0.706  0.704   
205  0.800  0.807  0.812  0.817  0.817  0.826  ...  0.873  0.877  0.879   
206  0.608  0.611  0.617  0.622  0.624  0.631  ...  0.697  0.703  0.708   
207    NaN    NaN    NaN    NaN    NaN    NaN  ...    NaN    NaN    NaN   

      2013   2014   2015   2016   2017   2018   2019  
188  0.557  0.561  0.569  0.571  0.578  0.582  0.584  
189  0.537  0.547  0.553  0.558  0.563  0.569  0.571  
190    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
191  0.878  0.882  0.886  0.888  0.890  0.892  0.898  
192  0.727  0.733  0.738  0.743  0.746  0.750  0.753  
193  0.599  0.608  0.616  0.625  0.630  0.634  0.631  
194  0.490  0.496  0.499  0.501  0.505  0.507  0.513  
195  0.663  0.669  0.674  0.680  0.683  0.686  0.689  
196    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
197  0.688  0.691  0.695  0.699  0.701  0.703  0.705  
198  0.714  0.721  0.727  0.733  0.737  0.741  0.747  
199  0.759  0.766  0.770  0.772  0.776  0.779  0.791  
200  0.748  0.752  0.754  0.756  0.758  0.759  0.766  
201  0.607  0.617  0.624  0.634  0.639  0

From row `189` to `205`, the rows must be in anothe DataFrame, because it's not interesting for us to have it here (it will not fit with analysis). But it will be useful to have comparison data.

There are missing values for Marshall Islands, South Sudan (created in 2010) and Turkmenistan. But it's still interesting to keep those values.

## Divide the DataFrame in 2

In [21]:
df1=df.iloc[1:190]
df2=df.iloc[190:207]

In [22]:
df1

Country   1990   1991   1992   1993   1994  \
1                            Afghanistan  0.302  0.307  0.316  0.312  0.307   
2                                Albania  0.650  0.631  0.615  0.618  0.624   
3                                Algeria  0.572  0.576  0.582  0.586  0.590   
4                                Andorra    NaN    NaN    NaN    NaN    NaN   
5                                 Angola    NaN    NaN    NaN    NaN    NaN   
..                                   ...    ...    ...    ...    ...    ...   
185   Venezuela (Bolivarian Republic of)  0.644  0.654  0.660  0.662  0.662   
186                             Viet Nam  0.483  0.493  0.504  0.514  0.525   
187                                Yemen  0.401  0.401  0.404  0.406  0.408   
188                               Zambia  0.421  0.417  0.416  0.419  0.414   
189                             Zimbabwe  0.478  0.481  0.467  0.463  0.460   

      1995   1996   1997   1998  ...   2010   2011   2012   2013   2014  \
1    0.331  0.335  0.339  0.344  ...  0.472  0.477  0.489  0.496  0.500   
2    0.637  0.646  0.645  0.655  ...  0.745  0.764  0.775  0.782  0.787   
3    0.595  0.602  0.611  0.621  ...  0.721  0.728  0.728  0.729  0.736   
4      NaN    NaN    NaN    NaN  ...  0.837  0.836  0.858  0.856  0.863   
5      NaN    NaN    NaN    NaN  ...  0.517  0.533  0.544  0.555  0.565   
..     ...    ...    ...    ...  ...    ...    ...    ...    ...    ...   
185  0.666  0.668  0.670  0.672  ...  0.757  0.769  0.772  0.777  0.775   
186  0.537  0.548  0.547  0.567  ...  0.661  0.671  0.676  0.681  0.683   
187  0.414  0.421  0.426  0.431  ...  0.506  0.506  0.504  0.509  0.502   
188  0.415  0.416  0.416  0.416  ...  0.527  0.534  0.549  0.557  0.561   
189  0.453  0.453  0.447  0.442  ...  0.482  0.499  0.525  0.537  0.547   

      2015   2016   2017   2018   2019  
1    0.500  0.502  0.506  0.509  0.511  
2    0.788  0.788  0.790  0.792  0.795  
3    0.740  0.743  0.745  0.746  0.748  
4    0.862  0.866  0.863  0.867  0.868  
5    0.572  0.578  0.582  0.582  0.581  
..     ...    ...    ...    ...    ...  
185  0.769  0.759  0.743  0.733  0.711  
186  0.688  0.693  0.696  0.700  0.704  
187  0.483  0.474  0.467  0.468  0.470  
188  0.569  0.571  0.578  0.582  0.584  
189  0.553  0.558  0.563  0.569  0.571  

[189 rows x 31 columns]

In [23]:
list_na={i:df1.iloc[i].isna().sum()/df1.shape[1] for i in range(len(df1.index)) if df1.iloc[i].isna().sum()/df1.shape[1]}

In [24]:
sorted_na = sorted(list_na.items(), key=lambda kv: kv[1])

In [25]:
sorted_na

[(180, 0.06451612903225806),
 (10, 0.16129032258064516),
 (15, 0.16129032258064516),
 (47, 0.16129032258064516),
 (105, 0.16129032258064516),
 (4, 0.2903225806451613),
 (97, 0.2903225806451613),
 (156, 0.2903225806451613),
 (3, 0.3225806451612903),
 (11, 0.3225806451612903),
 (21, 0.3225806451612903),
 (26, 0.3225806451612903),
 (28, 0.3225806451612903),
 (33, 0.3225806451612903),
 (37, 0.3225806451612903),
 (48, 0.3225806451612903),
 (53, 0.3225806451612903),
 (57, 0.3225806451612903),
 (63, 0.3225806451612903),
 (89, 0.3225806451612903),
 (99, 0.3225806451612903),
 (102, 0.3225806451612903),
 (112, 0.3225806451612903),
 (126, 0.3225806451612903),
 (128, 0.3225806451612903),
 (130, 0.3225806451612903),
 (144, 0.3225806451612903),
 (145, 0.3225806451612903),
 (151, 0.3225806451612903),
 (169, 0.3225806451612903),
 (182, 0.3225806451612903),
 (67, 0.3870967741935484),
 (125, 0.41935483870967744),
 (131, 0.45161290322580644),
 (162, 0.45161290322580644),
 (5, 0.4838709677419355),
 (19, 0

In [26]:
#interpolation works only with numeric columns. So, select numeric columns first
data_numeric=df1._get_numeric_data().copy()

#select numeric columns with missing data
cols_interp=data_numeric.columns[data_numeric.isnull().sum()>0]
cols_interp

Index(['1990', '1991', '1992', '1993', '1994', '1995', '1996', '1997', '1998',
       '1999', '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007',
       '2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016'],
      dtype='object')

In [32]:
#interpolate them
data_numeric[cols_interp]=data_numeric[cols_interp].interpolate(method='linear')

data_numeric.isna().sum().sum()

df1[cols_interp]=data_numeric[cols_interp].interpolate(method='linear')

df1.columns[df1.isna().sum()>0]

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\frame.py:2963: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]


Index([], dtype='object')

In [28]:
df1

Country    1990     1991   1992     1993  \
1                            Afghanistan  0.3020  0.30700  0.316  0.31200   
2                                Albania  0.6500  0.63100  0.615  0.61800   
3                                Algeria  0.5720  0.57600  0.582  0.58600   
4                                Andorra  0.6085  0.61325  0.619  0.62325   
5                                 Angola  0.6450  0.65050  0.656  0.66050   
..                                   ...     ...      ...    ...      ...   
185   Venezuela (Bolivarian Republic of)  0.6440  0.65400  0.660  0.66200   
186                             Viet Nam  0.4830  0.49300  0.504  0.51400   
187                                Yemen  0.4010  0.40100  0.404  0.40600   
188                               Zambia  0.4210  0.41700  0.416  0.41900   
189                             Zimbabwe  0.4780  0.48100  0.467  0.46300   

        1994    1995    1996    1997    1998  ...   2010   2011   2012   2013  \
1    0.30700  0.3310  0.3350  0.3390  0.3440  ...  0.472  0.477  0.489  0.496   
2    0.62400  0.6370  0.6460  0.6450  0.6550  ...  0.745  0.764  0.775  0.782   
3    0.59000  0.5950  0.6020  0.6110  0.6210  ...  0.721  0.728  0.728  0.729   
4    0.62725  0.6315  0.6385  0.6475  0.6565  ...  0.837  0.836  0.858  0.856   
5    0.66450  0.6680  0.6750  0.6840  0.6920  ...  0.517  0.533  0.544  0.555   
..       ...     ...     ...     ...     ...  ...    ...    ...    ...    ...   
185  0.66200  0.6660  0.6680  0.6700  0.6720  ...  0.757  0.769  0.772  0.777   
186  0.52500  0.5370  0.5480  0.5470  0.5670  ...  0.661  0.671  0.676  0.681   
187  0.40800  0.4140  0.4210  0.4260  0.4310  ...  0.506  0.506  0.504  0.509   
188  0.41400  0.4150  0.4160  0.4160  0.4160  ...  0.527  0.534  0.549  0.557   
189  0.46000  0.4530  0.4530  0.4470  0.4420  ...  0.482  0.499  0.525  0.537   

      2014   2015   2016   2017   2018   2019  
1    0.500  0.500  0.502  0.506  0.509  0.511  
2    0.787  0.788  0.788  0.790  0.792  0.795  
3    0.736  0.740  0.743  0.745  0.746  0.748  
4    0.863  0.862  0.866  0.863  0.867  0.868  
5    0.565  0.572  0.578  0.582  0.582  0.581  
..     ...    ...    ...    ...    ...    ...  
185  0.775  0.769  0.759  0.743  0.733  0.711  
186  0.683  0.688  0.693  0.696  0.700  0.704  
187  0.502  0.483  0.474  0.467  0.468  0.470  
188  0.561  0.569  0.571  0.578  0.582  0.584  
189  0.547  0.553  0.558  0.563  0.569  0.571  

[189 rows x 31 columns]

In [30]:
# Save it into a csv with no index
df1.to_csv("HDI_cleaned.csv", index=False)

## Move the HDI Rank to the end

In [64]:
df2

Country   1990   1991   1992  \
190                                  Human Development    NaN    NaN    NaN   
191                        Very high human development  0.779  0.782  0.780   
192                             High human development  0.568  0.573  0.578   
193                           Medium human development  0.437  0.439  0.445   
194                              Low human development  0.352  0.353  0.355   
195                               Developing Countries  0.516  0.520  0.525   
196                                            Regions    NaN    NaN    NaN   
197                                        Arab States  0.556  0.559  0.565   
198                          East Asia and the Pacific  0.519  0.525  0.534   
199                            Europe and Central Asia  0.652  0.650  0.646   
200                    Latin America and the Caribbean  0.628  0.634  0.639   
201                                         South Asia  0.441  0.447  0.454   
202                                 Sub-Saharan Africa  0.402  0.404  0.404   
203                          Least Developed Countries  0.350  0.353  0.354   
204                     Small Island Developing States  0.595  0.598  0.603   
205  Organization for Economic Co-operation and Dev...  0.785  0.790  0.788   
206                                              World  0.598  0.601  0.601   

      1993   1994   1995   1996   1997   1998  ...   2010   2011   2012  \
190    NaN    NaN    NaN    NaN    NaN    NaN  ...    NaN    NaN    NaN   
191  0.790  0.795  0.799  0.804  0.804  0.813  ...  0.866  0.871  0.874   
192  0.584  0.588  0.596  0.604  0.610  0.616  ...  0.706  0.713  0.720   
193  0.451  0.457  0.464  0.471  0.476  0.484  ...  0.575  0.584  0.593   
194  0.356  0.356  0.361  0.368  0.373  0.379  ...  0.473  0.479  0.484   
195  0.530  0.534  0.541  0.547  0.553  0.559  ...  0.642  0.650  0.657   
196    NaN    NaN    NaN    NaN    NaN    NaN  ...    NaN    NaN    NaN   
197  0.571  0.576  0.581  0.587  0.594  0.599  ...  0.676  0.681  0.687   
198  0.542  0.549  0.559  0.568  0.576  0.581  ...  0.691  0.700  0.707   
199  0.643  0.638  0.642  0.646  0.651  0.657  ...  0.735  0.744  0.750   
200  0.644  0.650  0.656  0.662  0.669  0.675  ...  0.731  0.737  0.740   
201  0.460  0.466  0.473  0.480  0.486  0.492  ...  0.585  0.593  0.601   
202  0.406  0.407  0.411  0.415  0.418  0.422  ...  0.498  0.505  0.512   
203  0.358  0.358  0.366  0.374  0.381  0.388  ...  0.485  0.493  0.499   
204  0.608  0.612  0.618  0.624  0.629  0.632  ...  0.702  0.706  0.704   
205  0.800  0.807  0.812  0.817  0.817  0.826  ...  0.873  0.877  0.879   
206  0.608  0.611  0.617  0.622  0.624  0.631  ...  0.697  0.703  0.708   

      2013   2014   2015   2016   2017   2018   2019  
190    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
191  0.878  0.882  0.886  0.888  0.890  0.892  0.898  
192  0.727  0.733  0.738  0.743  0.746  0.750  0.753  
193  0.599  0.608  0.616  0.625  0.630  0.634  0.631  
194  0.490  0.496  0.499  0.501  0.505  0.507  0.513  
195  0.663  0.669  0.674  0.680  0.683  0.686  0.689  
196    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
197  0.688  0.691  0.695  0.699  0.701  0.703  0.705  
198  0.714  0.721  0.727  0.733  0.737  0.741  0.747  
199  0.759  0.766  0.770  0.772  0.776  0.779  0.791  
200  0.748  0.752  0.754  0.756  0.758  0.759  0.766  
201  0.607  0.617  0.624  0.634  0.639  0.642  0.641  
202  0.521  0.527  0.532  0.535  0.539  0.541  0.547  
203  0.504  0.510  0.516  0.520  0.525  0.528  0.538  
204  0.708  0.712  0.717  0.719  0.722  0.723  0.728  
205  0.883  0.886  0.889  0.892  0.894  0.895  0.900  
206  0.713  0.718  0.722  0.727  0.729  0.731  0.737  

[17 rows x 31 columns]

In [31]:
# Save it into a csv with no index
df2.to_csv("HDI_global_cleaned.csv", index=False)